libcloud/compute/drivers/scaleway.py

```diff
@@ -0,0 +1,662 @@
+        headers['Content-Type'] = 'application/json'
+        return headers
+
+
+def _to_lib_size(size):
+    return int(size / 1000 / 1000 / 1000)
+
+
+def _to_api_size(size):
+    return int(size * 1000 * 1000 * 1000)
+
+
+class ScalewayNodeDriver(NodeDriver):
+    """
+    Scaleway Node Driver Class
+
+    This is the primary driver for interacting with Scaleway.  It contains all
+    of the standard libcloud methods that Scaleway's API supports.
+    """
+
+    type = Provider.SCALEWAY
+    connectionCls = ScalewayConnection
+    name = 'Scaleway'
+    website = 'https://www.scaleway.com/'
+
+    SNAPSHOT_STATE_MAP = {
+        'snapshotting': VolumeSnapshotState.CREATING,
+        'available': VolumeSnapshotState.AVAILABLE,
+        'error': VolumeSnapshotState.ERROR
+    }
+
+    def list_locations(self):
+        """
+        List data centers available.
+
+        :return: list of node location objects
+        :rtype: ``list`` of :class:`.NodeLocation`
+        """
+        return [NodeLocation(driver=self, **copy.deepcopy(location))
+                for location in SCALEWAY_LOCATION_DATA]
+
+    def list_sizes(self, region=None):
+        """
+        List available VM sizes.
+
+        :param region: The region in which to list sizes
+        (if None, use default region specified in __init__)
+        :type region: :class:`.NodeLocation`
+
+        :return: list of node size objects
+        :rtype: ``list`` of :class:`.NodeSize`
+        """
+        response = self.connection._request_paged('/products/servers',
+                                                  region=region)
+        sizes = response['servers']
+
+        response = self.connection._request_paged(
+            '/products/servers/availability', region=region)
+        availability = response['servers']
+
+        return sorted([self._to_size(name, sizes[name], availability[name])
+                       for name in sizes], key=lambda x: x.name)
+
+    def _to_size(self, name, size, availability):
+        min_disk = (_to_lib_size(size['volumes_constraint']['min_size'] or 0)
+                    if size['volumes_constraint'] else 25)
+        max_disk = (_to_lib_size(size['volumes_constraint']['max_size'] or 0)
+                    if size['volumes_constraint'] else min_disk)
+
+        extra = {
+            'cores': size['ncpus'],
+            'monthly': size['monthly_price'],
+            'arch': size['arch'],
+            'baremetal': size['baremetal'],
+            'availability': availability['availability'],
+            'max_disk': max_disk,
+            'internal_bandwidth': int(
+                (size['network']['sum_internal_bandwidth'] or 0) /
+                (1024 * 1024)),
+            'ipv6': size['network']['ipv6_support'],
+            'alt_names': size['alt_names'],
+        }
+
+        return NodeSize(id=name,
+                        name=name,
+                        ram=int((size['ram'] or 0) / (1024 * 1024)),
```

Comment:
Let's not worry about this, but I think this is going to return an integer in Python 2, and a more precise float in Python 3, since we only have ints here. Using 1024.0 would have avoided the issue.

Reply:
Pretty sure the `int()` wrapping it will return an `int` in any case?

Reply:
Right sorry I was thinking the int() was on `size['ram']`, not the whole expression. I guess using // could have been clearer to me but that's quite an obscure operator so you made the right choice, thanks!